contributors/design-proposals/secure_watch.md

```diff
@@ -0,0 +1,208 @@
+
+ The crucial requirements here are:
+
+1. Whenever "list" request returns a list of objects and a resource version "rv",
+starting a watch from the returned "rv" will never drop any events.
+2. For a given watch request (with resource version "rv"), the returned stream
+of events is always the same (e.g. very slow laggin watch may not cause dropped
+events). 
+
+ To satisfy the above requirements, we can't really rely only on the existing
+tools/frameworks that we have. Theoretically, we would be able to build the
+mapping from an object to list of nodes it is bound to using standard
+reflector/informer framework. But having separate informers for different kinds
+of objects (which is the only possible way to use them) may result in races.
+As an example, imagine a pod P referencing secret S, bounding pod P to node N
+happening right before creating secret S. Then the following race may potentially
+happen:
+1. user observes binding of pod P, he tries to retrieve the current value of a
+secret and it still doesn't exist. From now she knows that as soon as it will
+be created, it should be delivered via watch to him.
+2. however, pod informer is lagging in our code, and we first observe secret
+creation. It's not referenced yet (we didn't yet observe the pod), so we don't
+send any event.
+3. we observe the pod binding, but according to our semantics, we don't send
+any event.
+And we didn't send the expected "add secret" event to the watcher.
+
+ To solve the problem reliably, we need to be able to correctly serialize the
+watch events between different object types in a deterministic way.
+
+ One potential solution would be to identify this watch by a resource version
+combined from resource versions coming from different object kinds (e.g.
+pods have rv = rv1, secrets have rv = rv2, ...). Then we could keep the history
+of objects being processed to update the in-memory mapping and that would be
+deterministic. But the order might have been different e.g. after restarting
+apiserver. Which also means that it wouldn't work in HA setups (assuming we
+don't have some external "serializing" component).
+
+ Fortunately, we can solve it in much simpler way, with one additional assumption:
+
+1. All object types necessary to determine the in-memory mapping share the same
+resource version series.
+
+ With that assumption, we can have a "multi-object-type" watch that will serialize
+events for different object types for us. Having exactly watch responsible for
+delivering all objects (pods, secrets, ...) will guarantee that if we are currently
+at resource version rv, we procesed object of all types up to rv and nothing with
+resource version greater than rv. Which is exactly what we need.
+
+NOTE: We are not going to implement a generic "multi-object-type" watch and
+expose it in the api (which is much bigger task). This will be purely implementation
+detail hidden in the code (see more details in the next section).
+
+
+### Implementation details
```

Comment:
I think that we can't really drop the assumptions that RVs of objects returned via watch are in non-decreasing order, because it would break __all__ the machinery we have.

That said, one potential thing that comes to my mind is to say that we deliver all those ADD and DELETE events being result of pod/pv/.. modification (that I described that we don't deliver) but with the following caveat:
1. if a pod change in rv=X (or any other object) is causing add or delete of secrets (or anything else) s1, s2, s3, then we also deliver them as watch events, but all of them with the same RV=X (equal to the one of changed pod).

However, that has two main implications:
1. we end up with objects that has RVs not being their real RVs (though I don't think this will have any bad implications)
2. we end up with (potentially) multiple watch events with the same RV. This is a problem if the watch breaks/closes in between of sending them.

A workaround then would be to resend all of them when watch is renewed.
So when we get a request with rv=x, instead of starting from x, we start the watch by sending all events from x-1, and then user needs to protect himself for processing re-adding already added object or deleting the non-existing one. Though in thing like ThreadSafeStore it would just work.
But that is again change in the semantics.


Another potential modification to the above would be, instead of having a watch event to be a secret (or any other object), change it to actually be list of secrets. Then the above problem disappears, but we end up with a completely different API. So then it probably can't be extension to existing WATCH, but should be some different API endpoint (like WATCHLIST, though this name has already been used for a different purpose :)).  And we end up in a situation when our machinery (e.g. reflector) again doesn't work (it's possible to make it work, but it requires work).

@deads2k WDYT?